### PR TITLE
feat: sort by labels in shopping list copy if labels toggled

### LINKED ISSUE #3227: [BUG] - Copying shoppings lists results in blank output
### First Check

- [X] This is not a feature request.
- [X] I added a very descriptive title to this issue (title field is above this).
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Mealie documentation, with the integrated search.
- [X] I already read the docs and didn't find an answer.
- [X] This issue can be replicated on the demo site (https://demo.mealie.io/).

### What is the issue you are experiencing?

When copying shopping list items (either as text or markdown), each line in the output is just a blank string. Having had a look at the `copyListItems` function, it uses the `display` property on each list item to populate the copy output. The display property seems to always come back as blank in the `nightly` build of Mealie but *not* in the `latest` image, so it must be a recent regression. Looks like a change has been made on the backend side which is breaking this display property.

### Steps to Reproduce

Go to the demo mealie site and log in. Go to [any shopping list](https://demo.mealie.io/shopping-lists/5ad01392-93aa-4fd9-939d-5fc7af1a4fc8) with some items and click "Copy as text". The output will be a set of blank lines equal to the number of items in the list.

### Please provide relevant logs

No logs, UI only issue

### Mealie Version

Nightly, https://github.com/mealie-recipes/mealie/commit/5ac7645350ba580f212fbd1c0d29eb96c1317076

### Deployment

Docker (Linux)

### Additional Deployment Details

Docker compose on Linux with Postgres DB

### PR DESCRIPTION
**Pull request by ollywelch**
## What type of PR is this?

- feature

## What this PR does / why we need it:

This is a simple quality of life feature when copying shopping lists. Currently the lists always get copied in the order that is displayed when label sorting is off. So, when you toggle label sorting on, the copy function ends up copying in a different order to what is displayed. This change deals with the case when label sorting is on.

### Updates
- Updated the `copyListItems` function to unpack the values of the `itemsByLabel` mapping into the `items` to copy if label sorting is toggled on. Otherwise, it will have the same behaviour as before. 

## Which issue(s) this PR fixes:

Fixes https://github.com/mealie-recipes/mealie/issues/3227

## Special notes for your reviewer:

Go easy, not touched Vue before now :)

Also not sure if I need an issue creating for this, can create one and amend my commit msg to include it if needed.

## Testing

Ran the UI, made a shopping list with labels, then checked the copy output with label sorting toggled on and off.

### CODE REVIEW

**Comment by ollywelch** [comment]
Actually discovered another issue while testing this, looks like the copy function is using the "display" property of the shopping list item which is always coming back blank - this is only an issue in the nightly build and not latest so must be a recent API change

**Comment by ollywelch** [comment]
Raised https://github.com/mealie-recipes/mealie/issues/3227, will see if I can also fix this in my PR

**Comment by michael-genson** [comment]
I haven't had the chance to fully test it out yet, but generally looks good! I went ahead and fixed that `post_validate` issue by renaming all the `post_validate` methods (I probably should've done that originally).

I should be able to take a closer look at this later today

**Comment by michael-genson** [comment]
Would it make sense to also add the labels into the copied shopping list? I feel like if you want to keep them in label order, you probably also want the labels. Something like:

[Produce]
apples
bananas

[Meat]
1 lb chicken thighs

etc. I think we should at least have it in the markdown copy mode, though I'd argue you'd want it in the plaintext one too. I'm thinking something like this: https://github.com/mealie-recipes/mealie/pull/3032

**Comment by ollywelch** [comment]
Nice one with just renaming the methods, that's a bit neater than calling the same `post_validate` method from the super class :+1: 

With copying the headings, I think you're right. I've had to re-engineer it and not use the `copyMarkdownCheckList` function, rather just use the `copyPlain` function regardless and do the markdown formatting up front. But have tested and it works well

File: frontend/pages/shopping-lists/_id.vue [hunk 1]
```diff
@@ -337,21 +337,46 @@ export default defineComponent({
...
+      const text: string[] = [];
 
-      if (!items) {
-        return;
+      if (preferences.value.viewByLabel) {
+        // if we're sorting by label, we want the copied text in subsections
+        Object.entries(itemsByLabel.value).forEach(([label, items]) => {
+
+          // add an appropriate heading for the label depending on the copy format
+          text.push(formatCopiedLabelHeading(copyType, label))
```
Comment (michael-genson) [inline frontend/pages/shopping-lists/_id.vue:351]: Can we add an empty line before each header? Unless it's the first header. So it renders like this:

```
[Label 1]
- [ ] item 1
- [ ] item 2

[Label 2]
- [ ] item 3
- [ ] item 4
```

But skip the first index, so something like:

```js
Object.entries(itemsByLabel.value).forEach(([label, items], idx) => {
  if (idx) {
    text.push("");
  }
  ...
}
```

**Comment by michael-genson** [comment]
One small change and I think this is good to go. Thanks!

**Review by michael-genson — CHANGES_REQUESTED** [review]
Sorry, I just realized I left this as pending!

File: frontend/pages/shopping-lists/_id.vue (hunk 1 above)
Comment (ollywelch) [reply frontend/pages/shopping-lists/_id.vue:351]: Just made the update, I initially tried this yesterday but couldn't understand why I was getting two blank lines by pushing `"\n"` to the array :facepalm: 

Thanks for the guidance :smile:

**Review by michael-genson — APPROVED** [review]
Looks great, thank you!

**Pull request merged**